design/20241124-rename-bunde-to-clusterbundle.md

```diff
@@ -0,0 +1,252 @@
+otherwise rely on the Kubernetes garbage collector to delete orphans. But this should be tested/verified.
+
+### Temporary Technical Controller
+
+When the functional controller is migrated to target `ClusterBundle`, we want to introduce a temporary technical
+controller targeting the now deprecated `Bundle` resource.
+
+This controller should just create or update the `ClusterBundle` with the same name as the reconciled `Bundle`.
+Since the APIs should be quite similar, this should be a simple controller.
+
+- `Bundle` `.spec` should be copied (and eventually slightly converted) into `ClusterBundle` `.spec`
+- `Bundle` `.status` should be updated from `ClusterBundle` `.status`
+
+To allow a user to migrate their resources from `Bundle` to `ClusterBundle`, the controller should not act
+on `Bundle` delete events - nor should it add an owner reference to `ClusterBundle` to avoid the Kubernetes garbage collector
+acting on deletes.
+
+For reconciling a `ClusterBundle` from `Bundle` we should investigate the possibility of "smart field manager name logic".
+It should be possible to read the field manager name used (by user) to create/update the `Bundle` resource.
+If we use the same field manager name in controller (per resource) to create/update `ClusterBundle`, we think it's less
+likely that users will have to force conflicts when migrating from `Bundle` to `ClusterBundle`.
+
+### Proposed API changes when migrating
+
+Improving the trust-manager API while migrating is not something we have to do, but
+appears like nice opportunity to fix some of the long-standing proposed API changes.
+
+There is an [open draft PR for proposed (and breaking) API changes](https://github.com/cert-manager/trust-manager/pull/486),
+but in order to make this a non-braking change for users, only the `ClusterBundle` API will include these changes.
+
+While this is subject to change, an example on how a user would migrate from `Bundle` to `ClusterBundle` without
+changing anything for target resources is shown below:
+
+```yaml
+---
+apiVersion: trust.cert-manager.io/v1alpha1
+kind: Bundle
+metadata:
+  name: my-bundle
+spec:
+  sources:
+    - configMap:
+        name: my-ca
+        key: ca.crt
+  target:
+    configMap:
+      key: ca.crt
+    additionalFormats:
+      jks:
+        key: bundle.jks
+      pkcs12:
+        key: bundle.p12
+```
+
+Which would be migrated to:
+
+```yaml
+---
+apiVersion: trust-manager.io/v1alpha1
+kind: ClusterBundle
+metadata:
+  name: my-bundle
+spec:
+  sources:
+    - configMap:
+        name: my-ca
+        key: ca.crt
+  target:
+    configMap:
+      - key: ca.crt
+      - key: bundle.jks
+        format: JKS
+      - key: bundle.p12
+        format: PKCS12
+    namespaceSelector: {} # Selecting all namespaces here, but now required in API
+```
+
+### Risks and Mitigations
+
+#### Target configmaps/secrets are accidentally deleted
```

Comment:
There are two conflicting goals here:
- we do not want to break the behavior when deleting the `Bundle` resource
- we want to support users who remove the `Bundle` resource and start using the `ClusterBundle` instead.

Reply:
An alternative approach here would be to add owner references and document how to remove those owner references when you want to start using the `ClusterBundle` as the new "source of truth".

Reply:
If "owner references" are too complex, we can also implement our own garbage collector (eg. using labels).

Reply:
Why do you consider this as "conflicting goals"? I did not mention "do not want to break the behavior when deleting the Bundle resource" as a goal specifically, did I? 🤔 Considering the alpha state of Bundle API, I don't see a big harm in potentially orphaning some resources that can easily be deleted by the user as a workaround. The alternative could be a lot more destructive.

As a user, I **don't** want my `ClusterBundle` to be deleted when the corresponding `Bundle` is deleted. Maybe I am missing your point. Please clarify!